### PR TITLE
Save and restore previous terminal after setting the terminal for checking if terminal supports colors

### PR DESCRIPTION
**Pull request by augusto2112**
The call to "set_curterm" inside the "terminalHasColors" function breaks
the EditLine configuration on some Linux distributions, causing certain
characters that have functions bound to them to not show up and
backspace to stop deleting characters (only visually). This patch
ensures that term struct is restored after the routine for cheking if
terminal supports colors is done, which fixes the aforementioned issue.

Reviewed By: labath

Differential Revision: https://reviews.llvm.org/D95230

(cherry picked from commit 075de2d8a7567a6a39e8477407960aa2545b68c2)

@adrian-prantl this is a patch I made back in March to fix an issue where certain characters wouldn't show up in the REPL on Linux, but I forgot to cherry-pick it from llvm.org 🤦‍♂️

### CODE REVIEW

**Comment by augusto2112** [comment]
@swift-ci test

**Review by adrian-prantl — APPROVED** [review]

**Comment by adrian-prantl** [comment]
Please wait for a review from @fredriss who is managing the swift/release/5.5 branch.

**Comment by augusto2112** [comment]
Build failure in swiftpm:

```
10:30:37 
/Users/buildnode/jenkins/workspace/apple-llvm-project-pr-macos/swiftpm/Sources/Commands/SwiftBuildTool.swift:86:35: error: argument passed to call that takes no arguments
10:30:37     @OptionGroup(_hiddenFromHelp: true)
```

**Comment by augusto2112** [comment]
@swift-ci test macOS

**Comment by augusto2112** [comment]
@swift-ci test macOS

**Comment by augusto2112** [comment]
@fredriss could you review this please?

**Comment by fredriss** [comment]
While this seems obviously correct, I don't see enough upside to justify taking such a change in generic code at this point of the release. It'll make it into the next stable release automatically.

**Pull request closed without merging**
